File: csharp/Assembler/TemplateLoader/LoaderNormal.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Assembler.TemplateCommon;

namespace Assembler.TemplateLoader;
// Always use Base JsonObject/JsonArray types for consistent processing
/// <summary>
/// Handles loading and caching of HTML templates from the file system
/// </summary>
public static class LoaderNormal
{
    #region Loading Templates

    private static readonly ConcurrentDictionary<string, Dictionary<string, (string html, string? json)>> _htmlTemplatesCache = new();

    /// <summary>
    /// Loads HTML files and corresponding JSON files from the specified application site directory, caching the output per appSite
    /// </summary>
    public static Dictionary<string, (string html, string? json)> LoadGetTemplateFiles(string rootDirPath, string appSite)
    {
        Logger.Debug($"LoadGetTemplateFiles called for appSite: {appSite}", "LoaderNormal");

        var cacheKey = Path.GetDirectoryName(rootDirPath) + "|" + appSite;
        if (_htmlTemplatesCache.TryGetValue(cacheKey, out var cached))
        {
            Logger.Debug($"Returning cached templates for {appSite} ({cached.Count} templates)", "LoaderNormal");
            return cached;
        }

        var result = new Dictionary<string, (string html, string? json)>();
        var appSitesPath = Path.Combine(rootDirPath, "AppSites", appSite);

        if (!Directory.Exists(appSitesPath))
        {
            Logger.Warn($"AppSites directory not found: {appSitesPath}", "LoaderNormal");
            return result;
        }

        Logger.Debug($"Loading templates from: {appSitesPath}", "LoaderNormal");

        foreach (var file in Directory.GetFiles(appSitesPath, "*.html", SearchOption.AllDirectories))
        {
            var fileName = Path.GetFileNameWithoutExtension(file);
            var key = ($"{appSite.ToLowerInvariant()}_{fileName.ToLowerInvariant()}");
            var htmlContent = File.ReadAllText(file);

            Logger.Debug($"Loading template: {key} (html size: {htmlContent.Length})", "LoaderNormal");

            // Find JSON file case-insensitively
            var jsonFile = Path.ChangeExtension(file, ".json");
            string? jsonContent = null;

            // Try exact match first
            if (File.Exists(jsonFile))
            {
                jsonContent = File.ReadAllText(jsonFile);
                Logger.Debug($"Found JSON file for {key} (size: {jsonContent.Length})", "LoaderNormal");
            }
            else
            {
                // Try case-insensitive search in the same directory
                var directory = Path.GetDirectoryName(file);
                var baseFileName = Path.GetFileNameWithoutExtension(file);
                if (!string.IsNullOrEmpty(directory))
                {
                    var jsonFiles = Directory.GetFiles(directory, "*.json", SearchOption.TopDirectoryOnly);
                    var matchingJson = jsonFiles.FirstOrDefault(f =>
                        string.Equals(Path.GetFileNameWithoutExtension(f), baseFileName, StringComparison.OrdinalIgnoreCase));

                    if (matchingJson != null)
                    {
                        jsonContent = File.ReadAllText(matchingJson);
                        Logger.Debug($"Found JSON file (case-insensitive) for {key} (size: {jsonContent.Length})", "LoaderNormal");
                    }
                }
            }

            result[key] = (htmlContent, jsonContent);
        }

        Logger.Info($"Loaded {result.Count} templates for {appSite}", "LoaderNormal");

        _htmlTemplatesCache.TryAdd(cacheKey, result);
        return result;
    }

    /// <summary>
    /// Clear all cached templates (useful for testing or when templates change)
    /// </summary>
    public static void ClearCache()
    {
        _htmlTemplatesCache.Clear();
    }

    #endregion
}

File: php/Assembler/src/TemplateLoader/LoaderNormal.php
<?php

namespace Assembler\TemplateLoader;

use Assembler\TemplateCommon\Logger;

/**
 * Template result structure
 */
class TemplateResult
{
    public string $html;
    public ?string $json;

    public function __construct(string $html, ?string $json = null)
    {
        $this->html = $html;
        $this->json = $json;
    }
}

/**
 * Handles loading and caching of HTML templates from the file system
 */
class LoaderNormal
{
    private static array $htmlTemplatesCache = [];

    /**
     * Loads HTML files and corresponding JSON files from the specified application site directory, caching the output per appSite
     * @param string $rootDirPath Root directory path
     * @param string $appSite Application site name
     * @return array<string, TemplateResult> Array of templates
     */
    public static function loadGetTemplateFiles(string $rootDirPath, string $appSite): array
    {
        Logger::debug("LoadGetTemplateFiles called for appSite: $appSite", 'LoaderNormal');

        $cacheKey = dirname($rootDirPath) . '|' . $appSite;

        if (isset(self::$htmlTemplatesCache[$cacheKey])) {
            $cached = self::$htmlTemplatesCache[$cacheKey];
            Logger::debug("Returning cached templates for $appSite (" . count($cached) . " templates)", 'LoaderNormal');
            return $cached;
        }

        $result = [];
        $appSitesPath = $rootDirPath . DIRECTORY_SEPARATOR . 'AppSites' . DIRECTORY_SEPARATOR . $appSite;

        if (!is_dir($appSitesPath)) {
            Logger::warn("AppSites directory not found: $appSitesPath", 'LoaderNormal');
            self::$htmlTemplatesCache[$cacheKey] = $result;
            return $result;
        }

        Logger::debug("Loading templates from: $appSitesPath", 'LoaderNormal');

        // Recursively find all HTML files
        $iterator = new \RecursiveIteratorIterator(
            new \RecursiveDirectoryIterator($appSitesPath, \RecursiveDirectoryIterator::SKIP_DOTS),
            \RecursiveIteratorIterator::LEAVES_ONLY
        );

        foreach ($iterator as $file) {
            if ($file->isFile() && $file->getExtension() === 'html') {
                $fileName = $file->getBasename('.html');
                $key = strtolower($appSite) . '_' . strtolower($fileName);

                $htmlContent = file_get_contents($file->getPathname());
                Logger::debug("Loading template: $key (html size: " . strlen($htmlContent ?: '') . ")", 'LoaderNormal');

                // Find JSON file case-insensitively
                $jsonFile = substr($file->getPathname(), 0, -5) . '.json'; // Replace .html with .json
                $jsonContent = null;

                if (file_exists($jsonFile)) {
                    $jsonContent = file_get_contents($jsonFile);
                    Logger::debug("Found JSON file for $key (size: " . strlen($jsonContent ?: '') . ")", 'LoaderNormal');
                } else {
                    // Try case-insensitive search in the same directory
                    $dir = dirname($file->getPathname());
                    $baseName = strtolower($file->getBasename('.html'));
                    $files = scandir($dir);

                    if ($files !== false) {
                        foreach ($files as $entry) {
                            if ($entry !== '.' && $entry !== '..' && is_file($dir . DIRECTORY_SEPARATOR . $entry)) {
                                if (strtolower(pathinfo($entry, PATHINFO_EXTENSION)) === 'json') {
                                    $entryBase = strtolower(pathinfo($entry, PATHINFO_FILENAME));
                                    if ($entryBase === $baseName) {
                                        $matchedJsonPath = $dir . DIRECTORY_SEPARATOR . $entry;
                                        $jsonContent = file_get_contents($matchedJsonPath);
                                        Logger::debug("Found JSON file (case-insensitive) for $key (size: " . strlen($jsonContent ?: '') . ")", 'LoaderNormal');
                                        break;
                                    }
                                }
                            }
                        }
                    }
                }

                $result[$key] = new TemplateResult($htmlContent ?: '', $jsonContent ?: null);
            }
        }

        Logger::info("Loaded " . count($result) . " templates for $appSite", 'LoaderNormal');

        self::$htmlTemplatesCache[$cacheKey] = $result;
        return $result;
    }

    /**
     * Clear all cached templates (useful for testing or when templates change)
     */
    public static function clearCache(): void
    {
        self::$htmlTemplatesCache = [];
    }
}
?>